MonoGame.Framework/Random.cs

```diff
@@ -0,0 +1,404 @@
+            /// <param name="min">Minimum possible value</param>
+            /// <param name="max">Maximum possible value</param>
+            /// <returns>Random single-precision floating point number</returns>
+            public float NextFloat(float min, float max)
+            {
+                return (float)NextDouble((double)min, (double)max);
+            }
+
+            #endregion
+
+            #endregion
+
+            #endregion
+
+            #region Private methods
+
+            private ulong XORShift64Star()
+            {
+                _state ^= _state >> 12;
+                _state ^= _state << 25;
+                _state ^= _state >> 27;
+                return unchecked(_state * 2685821657736338717UL);
+            }
+
+            private long Remap(long value, long oldMin, long oldMax, long newMin, long newMax)
+            {
+                return (((value - oldMin) * (newMax - newMin)) / (oldMax - oldMin)) + newMin;
+            }
+
+            private double Remap(double value, double oldMin, double oldMax, double newMin, double newMax) 
+            {
+                return (((value - oldMin) * (newMax - newMin)) / (oldMax - oldMin)) + newMin;
+            }
+
+            #region Integer values
+
+            private byte UInt64ToInt8(ulong uint64)
+            {
+                // Map the ulong onto a long, then remap that to a byte range by multiplying by the factor of 
+                // the maximum value of a byte divided by the maximum value of a long. However, if the long 
+                // is negative, return the same multiplied by -1.
+                if (UInt64ToInt64(uint64) < 0)
+                    return (byte)(-UInt64ToInt64(uint64) * ((double)byte.MaxValue / long.MaxValue));
+                return (byte)(UInt64ToInt64(uint64) * ((double)byte.MaxValue / long.MaxValue));
+            }
+
+            private short UInt64ToInt16(ulong uint64)
+            {
+                // Map the ulong onto a long, then remap that to a short range by multiplying by the factor of 
+                // the maximum value of a short divided by the maximum value of a long
+                return (short)(UInt64ToInt64(uint64) * ((double)short.MaxValue / long.MaxValue));
+            }
+
+            private int UInt64ToInt32(ulong uint64)
+            {
+                // Map the ulong onto a long, then remap that to an integer range by multiplying by the factor of 
+                // the maximum value of an integer divided by the maximum value of a long
+                return (int)(UInt64ToInt64(uint64) * ((double)int.MaxValue / long.MaxValue));
+            }
+
+            private long UInt64ToInt64(ulong uint64)
+            {
+                unsafe
+                {
+                    return (*(long*)&uint64) + long.MinValue;
+                }
+            }
+
+            #endregion
+
+            #region Floating point values
+
+            private float UInt64ToFloat(ulong uint64)
```

Comment:
UInt64ToInt64 is used inside of all of the integer Next() functions to filter the output of the xorshift to be an integer in the correct range so that it can be resized to a different integer type.